### PR TITLE
Support user resource name / displaytext with emoji, unicode chars, and some sql exception msg improvements

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR allows the below resource name / display text with unicode / emoji chars, and improves SQL exception msgs (doesn't include any sql stmts). Also, adds a validator for VM host/domain name.

- User VM
- Volume
- Volume Snapshot
- VM Snapshot
- Template
- ISO
- Disk Offering
- Compute Offering
- Network Offering
- VPC Offering
- Backup Offering
- Instance Group
- Affinity Group
- SSH Key
- Userdata
- Guest Network
- VPC
- Kubernetes Cluster
- Autoscale VM Group
- Account
- User
- Project
- Annotation

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Manually, tested the name / display names for these resources.

test_resource_names.py
[Checks support for emoji in the resource name for vm, volume, service & disk offering, template, iso, account(first/lastname)]

```
Test for deploy virtual machine ... === TestName: test_01_deploy_vm | Status : SUCCESS ===
ok
Test for create volume ... === TestName: test_02_create_volume | Status : SUCCESS ===
ok
Test for register template ... === TestName: test_03_register_template | Status : SUCCESS ===
ok
Test for register ISO ... === TestName: test_04_register_iso | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 4 tests in 201.090s

OK
```

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
Code looks good. Good to see tests

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@blueorangutan test alma8 kvm-alma8

**Comment by rohityadavcloud** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41910to42000.java [hunk 1]
```diff
@@ -80,4 +85,32 @@ public void updateSystemVmTemplates(Connection conn) {
...
+
+    private void checkAndUpdateAffinityGroupNameCharSetToUtf8mb4(Connection conn) {
+        logger.debug("Check and update char set for affinity group name to utf8mb4");
+        try {
+            PreparedStatement pstmt = conn.prepareStatement("SELECT MAX(LENGTH(name)) FROM `cloud`.`affinity_group`");
+            ResultSet rs = pstmt.executeQuery();
+            if (rs.next()) {
+                long maxLengthOfName = rs.getLong(1);
+                if (maxLengthOfName <= MAX_INDEXED_CHARS_IN_CHAR_SET_UTF8MB4) {
+                    pstmt = conn.prepareStatement("ALTER TABLE `cloud`.`affinity_group` MODIFY `name` VARCHAR(191) CHARACTER SET utf8mb4 NOT NULL");
```
Comment (harikrishna-patnala) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41910to42000.java:97]: @sureshanaparti please use the string constant for 191 in the prepareStatement, later someone may miss changing here.

**Review by harikrishna-patnala — COMMENTED** [review]
very minor comment @sureshanaparti , otherwise LGTM

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41910to42000.java (hunk 1 above)
Comment (sureshanaparti) [reply engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41910to42000.java:97]: ok @harikrishna-patnala, will check/update

**Comment by sureshanaparti** [comment]
> [SF] Trillian test result (tid-10914) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 58597 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr9340-t10914-kvm-centos7.zip Smoke tests completed. 114 look OK, 24 have errors, 0 did not run 

these tests fail on health check PR as well, [here](https://github.com/apache/cloudstack/pull/9316#issuecomment-2241008671). seems some issue with kvm-centos7 tests on main, not related to the changes in this PR.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by harikrishna-patnala — APPROVED** [review]
code LGTM

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 2]
```diff
@@ -437,9 +437,11 @@ public List<T> searchIncludingRemoved(SearchCriteria<T> sc, final Filter filter,
             }
             return result;
         } catch (final SQLException e) {
-            throw new CloudRuntimeException("DB Exception on: " + pstmt, e);
+            logger.error("DB Exception on: " + pstmt, e);
```
Comment (hsato03) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:440]:
```suggestion
            logger.error("DB Exception on: {}", pstmt, e);
```
The parameterized log message can also be used in the other logs.

**Review by hsato03 — APPROVED** [review]
LGTM

File: engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql [hunk 3]
```diff
@@ -151,6 +151,55 @@ WHERE
...
+ALTER TABLE `cloud`.`vm_snapshots` MODIFY `description` VARCHAR(255) CHARACTER SET utf8mb4 DEFAULT NULL;
+ALTER TABLE `cloud`.`vm_template` MODIFY `name` VARCHAR(255) CHARACTER SET utf8mb4 NOT NULL;
+ALTER TABLE `cloud`.`vm_template` MODIFY `display_text` VARCHAR(4096) CHARACTER SET utf8mb4 DEFAULT NULL COMMENT 'Description text set by the admin for display purpose only';
+ALTER TABLE `cloud`.`volumes` MODIFY `name` VARCHAR(255) CHARACTER SET utf8mb4 DEFAULT NULL COMMENT 'A user specified name for the volume';
+ALTER TABLE `cloud`.`vpc` MODIFY `name` VARCHAR(255) CHARACTER SET utf8mb4 DEFAULT NULL COMMENT 'vpc name';
+ALTER TABLE `cloud`.`vpc` MODIFY `display_text` VARCHAR(255) CHARACTER SET utf8mb4 DEFAULT NULL COMMENT 'vpc display text';
+ALTER TABLE `cloud`.`vpc_offerings` MODIFY `name` VARCHAR(255) CHARACTER SET utf8mb4  DEFAULT NULL COMMENT 'vpc offering name';
+ALTER TABLE `cloud`.`vpc_offerings` MODIFY `unique_name` VARCHAR(64) CHARACTER SET utf8mb4  DEFAULT NULL COMMENT 'unique name of the vpc offering';
+ALTER TABLE `cloud`.`vpc_offerings` MODIFY `display_text` VARCHAR(255) CHARACTER SET utf8mb4  DEFAULT NULL COMMENT 'display text';
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql:205]: use of idem_potent_change_column in these lines

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@sureshanaparti pl check and advise on the PR.

File: engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql (hunk 3 above)
Comment (sureshanaparti) [reply engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql:205]: i think, this is not required. here, it's char set change and rest all specs are kept the same in these columns. will check if i can update it with procedure IDEMPOTENT_CHANGE_COLUMN_CHARSET.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java (hunk 2 above)
Comment (sureshanaparti) [reply framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:440]: can take this up later if required.

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql (hunk 3 above)
Comment (sureshanaparti) [reply engine/schema/src/main/resources/META-INF/db/schema-41910to42000.sql]: @DaanHoogland updated columns with procedure IDEMPOTENT_MODIFY_COLUMN_CHAR_SET

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan help

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 kvm-centos7

**Comment by rohityadavcloud** [comment]
@sureshanaparti EL7 packaging needs to be kicked manually; otherwise smoketests will not run (we no longer run el7 pkging by default with BO). I think we can merge this based on EL8 smoketests.

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Pull request merged**